File: src/game/multiplayer/presence.ts
import type { PlayerSave } from '../save/types'
import { OUTFIT_COSMETIC_SLOT_ID, PET_COSMETIC_SLOT_ID } from '../save/types'
import { getSession } from './auth'
import { getLocalBackend } from './client'
import type { ActivityPresence, PublicPlayerProfile } from './types'

export type PresenceInput = Omit<
  ActivityPresence,
  'userId' | 'username' | 'updatedAt' | 'expiresAt' | 'guildName'
>

/**
 * What a save says about the player, before a backend stamps identity onto it.
 *
 * Presence advertises a skill level, and Combat is the one every character has,
 * so it is the default; a save without it falls back to whichever skill is first.
 */
export function presenceInputFromSave(save: PlayerSave): PresenceInput {
  const combat = save.skills.find((skill) => skill.skillId === 'SKL-0001')
  const skill = combat ?? save.skills[0]
  return {
    appearance: save.appearance,
    locationId: save.currentLocationId,
    currentActivityId: save.currentActivityId,
    skillId: skill?.skillId ?? null,
    skillLevel: skill?.level ?? null,
    outfitCosmeticId: save.cosmetics.equipped[OUTFIT_COSMETIC_SLOT_ID] ?? null,
    mountCosmeticId: save.cosmetics.equipped[PET_COSMETIC_SLOT_ID] ?? null,
  }
}

export function publishActivityPresence(save: PlayerSave): ActivityPresence | null {
  const session = getSession()
  if (!session) return null
  return getLocalBackend().upsertPresence(session, presenceInputFromSave(save))
}

export function clearActivityPresence(): void {
  const session = getSession()
  if (!session) return
  getLocalBackend().clearPresence(session.userId)
}

export function listPeersAtActivity(
  locationId: string,
  activityId: string | null,
  excludeSelf = true,
): ActivityPresence[] {
  const session = getSession()
  const peers = getLocalBackend().listPresence({ locationId, activityId })
  if (!excludeSelf || !session) return peers
  return peers.filter((row) => row.userId !== session.userId)
}

export function listPeersAtLocation(locationId: string, excludeSelf = true): ActivityPresence[] {
  const session = getSession()
  const peers = getLocalBackend().listPresence({ locationId })
  if (!excludeSelf || !session) return peers
  return peers.filter((row) => row.userId !== session.userId)
}

export function getPublicProfile(userId: string): PublicPlayerProfile | null {
  return getLocalBackend().publicProfile(userId)
}

export function sendFriendRequest(targetUserId: string): { ok: true } | { ok: false; reason: string } {
  const session = getSession()
  if (!session) return { ok: false, reason: 'Sign in required.' }
  return getLocalBackend().sendFriendRequest(session.userId, targetUserId)
}

File: packages/ik_net/lib/src/presence.dart
import 'package:ik_rules/ik_rules.dart';

import 'types.dart';

/// The skill a presence row advertises.
///
/// Combat is the one every character has, so it is the default; a save without
/// it falls back to whatever skill comes first.
PresenceInput presenceFromSave(PlayerSave save) {
  final combat = save.skills.where((skill) => skill.skillId == 'SKL-0001').firstOrNull;
  final fallback = save.skills.isEmpty ? null : save.skills.first;
  final skill = combat ?? fallback;
  return PresenceInput(
    appearance: save.appearance,
    locationId: save.currentLocationId,
    currentActivityId: save.currentActivityId,
    skillId: skill?.skillId,
    skillLevel: skill?.level,
    outfitCosmeticId: save.cosmetics.equipped[outfitCosmeticSlotId],
    mountCosmeticId: save.cosmetics.equipped[petCosmeticSlotId],
  );
}

/// Stable Local-chat location key while anywhere on the Citadel sub-map.
String citadelChatLocationIdOf() => citadelChatLocationId;

String citadelLocationId() => citadelLocationIdValue;

String citadelLocalChannelKey() =>
    chatChannelKey(const ChatChannel.local(citadelChatLocationId));

/// The two boards the Citadel keeps, each opened the way a shop is.
enum CitadelHubTab { bounties, bazaar }

const Map<CitadelHubTab, String> citadelHubTabLabels = <CitadelHubTab, String>{
  CitadelHubTab.bounties: 'Hourly Bounties',
  CitadelHubTab.bazaar: 'Grand Bazaar',
};

/// Which boards stand at [locationId].
///
/// The Plaza holds the notice board and the Market District holds the Bazaar, so
/// each district has its own reason to be walked to. Anywhere else has neither.
List<CitadelHubTab> citadelHubTabsFor(String locationId) {
  if (locationId == citadelPlazaId || locationId == citadelLocationIdValue) {
    return const <CitadelHubTab>[CitadelHubTab.bounties];
  }
  if (locationId == citadelMarketId) return const <CitadelHubTab>[CitadelHubTab.bazaar];
  return const <CitadelHubTab>[];
}

/// The heading above those links, naming the district rather than the boards.
String citadelHubTitleFor(String locationId) =>
    locationId == citadelMarketId ? 'Market District' : 'Citadel Plaza';

/// What the Citadel tab shows above its visitor list.
class CitadelHubSummary {
  const CitadelHubSummary({
    required this.locationId,
    required this.chatChannel,
    required this.visitorCount,
    required this.note,
  });

  final String locationId;
  final String chatChannel;
  final int visitorCount;
  final String note;

  Map<String, Object?> toJson() => <String, Object?>{
    'locationId': locationId,
    'chatChannel': chatChannel,
    'visitorCount': visitorCount,
    'note': note,
  };
}

CitadelHubSummary citadelHubSummary(int visitorCount) => CitadelHubSummary(
  locationId: citadelLocationIdValue,
  chatChannel: citadelLocalChannelKey(),
  visitorCount: visitorCount,
  note: 'Shared Citadel hub. Local chat is one room across every Citadel district.',
);
